packages/repository-tests/src/transaction/transactions.suite.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/repository-tests
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Entity, juggler, model, property} from '@loopback/repository';
+import {TransactionRepository} from '@loopback/repository/src';
+import {expect, skipIf, toJSON} from '@loopback/testlab';
+import {Suite} from 'mocha';
+import {withCrudCtx} from '../helpers.repository-tests';
+import {
+  CrudConnectorFeatures,
+  CrudTestContext,
+  DataSourceOptions,
+  TransactionRepositoryCtor,
+} from '../types.repository-tests';
+
+// Core scenarios around creating new model instances and retrieving them back
+// Please keep this file short, put any advanced scenarios to other files
+export function transactionSuite(
+  dataSourceOptions: DataSourceOptions,
+  repositoryClass: TransactionRepositoryCtor,
+  connectorFeatures: CrudConnectorFeatures,
+) {
+  skipIf<[(this: Suite) => void], void>(
+    !connectorFeatures.supportsTransactions,
+    describe,
+    `transactions`,
+    () => {
+      @model()
+      class Product extends Entity {
+        @property({
+          type: connectorFeatures.idType,
+          id: true,
+          generated: true,
+          description: 'The unique identifier for a product',
+        })
+        id: number | string;
+
+        @property({type: 'string', required: true})
+        name: string;
+
+        constructor(data?: Partial<Product>) {
+          super(data);
+        }
+      }
+
+      describe('create-retrieve within transaction', () => {
+        let repo: TransactionRepository<Product, typeof Product.prototype.id>;
+        before(
+          withCrudCtx(async function setupRepository(ctx: CrudTestContext) {
+            repo = new repositoryClass(Product, ctx.dataSource);
+            await ctx.dataSource.automigrate(Product.name);
+          }),
+        );
+
+        it('retrieves a newly created model in a transaction', async () => {
+          const tx: juggler.Transaction = await repo.beginTransaction({
+            isolationLevel: 'READ COMMITTED',
+          });
+          const created = await repo.create(
+            {name: 'Pencil'},
+            {transaction: tx},
+          );
+          expect(created.toObject()).to.have.properties('id', 'name');
+          expect(created.id).to.be.ok();
+
+          // const found = await repo.findById(created.id);
+          await tx.commit();
```

Comment:
How do you envision this test to fail when the database and/or the connector does not support transactions, or implement them incorrectly?

It seems to me that this test only verifies that create+retrieve preserves the data when both commands are executed in the same transaction. It will pass for any repository that implements `beginTransaction` API, regardless of whether it actually enforces the transaction or not.

Let's find more robust test scenarios please.

Here are few scenarios that come to my mind:

1. Create a transaction, `create` a new model from that transaction, COMMIT the transaction, `findById` the created model outside of the transaction - model should be found. This is a kind of a smoke test, it passes even if transaction isolation is not implemented.

2. Create a transaction, `create` a new model from that transaction, ROLLBACK the transaction, `findById` the created model outside of the transaction - no model should be returned. This verifies rollback feature, it does not require transactions to be isolated.

3. Create a transaction, `create` a new model from that transaction. `findById` without any transaction - no model should be returned. `findById` from the same transaction - the created model should be returned. This verifies isolation of individual transactions.